modules/core/02-client/keeper/extended_keeper.go

```diff
@@ -64,15 +83,206 @@ func (ek ExtendedKeeper) UpdateClient(ctx sdk.Context, clientID string, header e
 			)
 		}
 
-		// TODO: ensure the header has a valid QC
+		// apply all verification rules to the header
+		err := ek.checkHeader(ctx, clientID, tmHeader)
 
-		// get hash of the tx that includes this header
-		txHash := tmhash.Sum(ctx.TxBytes())
+		// good header, timestamp it on the canonical chain indexer
+		if err == nil {
+			txHash := tmhash.Sum(ctx.TxBytes())                         // get hash of the tx that includes this header
+			ek.AfterHeaderWithValidCommit(ctx, txHash, tmHeader, false) // invoke hooks to notify ZoneConcierge to timestamp this header
+		}
 
-		// invoke hooks
-		ek.AfterHeaderWithValidCommit(ctx, txHash, tmHeader)
+		// The header has a QC but {is on a fork, its timestamp is not monotonic}, timestamp it on the fork indexer, and return to avoid freezing the light client
+		// These two errors only happen upon dishonest majority
+		if sdkerrors.IsOf(err, types.ErrForkedHeaderWithQC, types.ErrHeaderNonMonotonicTimestamp) {
+			ctx.Logger().Debug("received a header that has QC but is on a fork")
+			txHash := tmhash.Sum(ctx.TxBytes())
+			ek.AfterHeaderWithValidCommit(ctx, txHash, tmHeader, true)
+			return err
+		}
+
+		// Upon other errors (that happen under honest majority), return error without timestamping it
+		if err != nil {
+			return err
+		}
 	}
 
-	// TODO: inject our own verification rules
+	// the header has a valid QC and is extending canonical chain
+	// follow the original verification rules to update client state
 	return ek.Keeper.UpdateClient(ctx, clientID, header)
 }
+
+// checkHeader checks
+// - if a header is on fork,
+// - if a header has a valid QC or not, and
+// - other miscellaneous verification rules
+// It is essentially `CheckHeaderAndUpdateState` without updating the state
+// (adapted from https://github.com/cosmos/ibc-go/blob/v5.0.0/modules/core/02-client/keeper/client.go#L60)
+func (ek ExtendedKeeper) checkHeader(ctx sdk.Context, clientID string, tmHeader *ibctmtypes.Header) error {
+	clientState, found := ek.GetClientState(ctx, clientID)
+	if !found {
+		return sdkerrors.Wrapf(types.ErrClientNotFound, "cannot update client with ID %s", clientID)
+	}
+
+	clientStore := ek.ClientStore(ctx, clientID)
+	if status := clientState.Status(ctx, clientStore, ek.cdc); status != exported.Active {
+		return sdkerrors.Wrapf(types.ErrClientNotActive, "cannot update client (%s) with status %s", clientID, status)
+	}
+
+	// Check if the header is on a fork
+	// If the consensus state exists, and it does not match this header, then this header is on a fork
+	// (adapted from https://github.com/cosmos/ibc-go/blob/v5.0.0/modules/light-clients/07-tendermint/types/update.go#L63-L77)
+	isOnFork := false
+	conflictingState, _ := ibctmtypes.GetConsensusState(clientStore, ek.cdc, tmHeader.GetHeight())
```

Comment:
Why not inherit the name `prevConsState`? It might not be conflicting, is it?

Reply:
Nice catch! Was debugging and forgot to change it back 👍